File: questions/981.TimeBasedKey-ValueStore.ts
class TimeMap {
  map: Record<string, { value: string; timestamp: number }[]> = {}

  set(key: string, value: string, timestamp: number): void {
    this.map[key] ||= []
    this.map[key].push({ value, timestamp })
  }

  get(key: string, timestamp: number): string {
    if (!this.map[key]?.length) return ''

    const values = this.map[key]
    let result = ''
    let left = 0
    let right = values.length - 1

    while (left <= right) {
      const middle = Math.floor((left + right) / 2)

      if (values[middle].timestamp <= timestamp) {
        result = values[middle].value
        left = middle + 1
      } else {
        right = middle - 1
      }
    }

    return result
  }
}

const timeMap = new TimeMap()
timeMap.set('foo', 'bar', 1)
console.log(timeMap.get('foo', 1))
console.log(timeMap.get('foo', 3))

timeMap.set('foo', 'bar2', 4)
console.log(timeMap.get('foo', 4))
console.log(timeMap.get('foo', 5))

File: questions/981.TimeBasedKey-ValueStore.go
package questions

import "math"

type TimeMapValue struct {
	Value     string
	Timestamp int
}

type TimeMap struct {
	Map map[string][]*TimeMapValue
}

func (t *TimeMap) Set(key string, value string, timestamp int) {
	_, ok := t.Map[key]
	if !ok {
		t.Map[key] = []*TimeMapValue{}
	}
	t.Map[key] = append(t.Map[key], &TimeMapValue{
		Value:     value,
		Timestamp: timestamp,
	})
}

func (t *TimeMap) Get(key string, timestamp int) string {
	_, ok := t.Map[key]
	if !ok {
		return ""
	}
	if len(t.Map[key]) == 0 {
		return ""
	}

	values := t.Map[key]
	result := ""
	start := 0
	end := len(values) - 1

	for start <= end {
		middle := int(math.Floor(float64((start + end) / 2)))

		if values[middle].Timestamp <= timestamp {
			result = values[middle].Value
			start = middle + 1
		} else {
			end = middle - 1
		}
	}

	return result
}
